### PR TITLE
Fix issues with fifos blocking on open

### LINKED ISSUE #26561: deadlock in monitor.go
**Description**

Docker ps command hangs after the daemon process runs for a few days. Usually many containers has been killed and start on the host. I still cannot reproduce this problem manually. But it happens definitely after daemon runs for a few days. 
**Steps to reproduce the issue:**
1.
2.
3.

**Describe the results you received:**

**Describe the results you expected:**

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   8eab29e
 Built:        Thu Jul 28 23:54:00 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   8eab29e
 Built:        Thu Jul 28 23:54:00 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Client:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   8eab29e
 Built:        Thu Jul 28 23:54:00 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   8eab29e
 Built:        Thu Jul 28 23:54:00 2016
 OS/Arch:      linux/amd64
wanglu@mesos22.aws.dev:~$ sudo docker info
Containers: 12045
 Running: 29
 Paused: 0
 Stopped: 12016
Images: 50
Server Version: 1.12.0
Storage Driver: overlay2
 Backing Filesystem: extfs
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: null bridge host overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 4.5.0-0.bpo.2-amd64
Operating System: Debian GNU/Linux 8 (jessie)
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 14.69 GiB
Name: mesos22
ID: WFEI:YJJD:2XQ3:GQHW:YVQZ:UCPT:NSLJ:KEPN:TDKS:WDZ7:PLSV:IZJN
Docker Root Dir: /data/data/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

### LINKED ISSUE #27185: Docker Engine hangs after calling "docker stop"
**Description**
This issue is sporadic and does only occur once a day on our Jenkins node. We are running our Jobs in a Docker container using the Jenkins Docker pipeline plugin. At the end of a Job the plugin stops the container and will hang from there on.

**Steps to reproduce the issue:**
1. This is sporadic and not reproducible step by step

**Describe the results you received:**
No repsonse from daemon to "docker ps", "docker stop <container>", docker kill <container>"

**Describe the results you expected:**
Normal responses when calling "docker ps", ...

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 05:22:43 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 05:22:43 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 5
 Running: 1
 Paused: 0
 Stopped: 4
Images: 2
Server Version: 1.12.1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 29
 Dirperm1 Supported: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: host bridge null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: apparmor
Kernel Version: 3.13.0-48-generic
Operating System: Ubuntu 14.04.4 LTS
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 992.5 MiB
Name: jenkins-node-1.internal
ID: IOJZ:PGC4:4HVY:4DKY:32EP:...
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 23
 Goroutines: 45
 System Time: 2016-10-06T09:33:37.423801303Z
 EventsListeners: 0
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Docker is running on an EC2 t2.micro instance with Ubuntu 14.04.3.

Log output of a normal "docker stop" called from the Jenkins plugin:
```
time="2016-10-05T13:59:28.707836999Z" level=debug msg="Calling POST /v1.24/containers/75190c7ed862395af5a4dfed025f6f27a3bb39da9393a61ef724e698ab8d325a/stop?t=10" 
time="2016-10-05T13:59:28.707907143Z" level=debug msg="Sending 15 to 75190c7ed862395af5a4dfed025f6f27a3bb39da9393a61ef724e698ab8d325a" 
time="2016-10-05T13:59:38.708608718Z" level=info msg="Container 75190c7ed862395af5a4dfed025f6f27a3bb39da9393a61ef724e698ab8d325a failed to exit within 10 seconds of signal 15 - using the force" 
time="2016-10-05T13:59:38.708649191Z" level=debug msg="Sending 9 to 75190c7ed862395af5a4dfed025f6f27a3bb39da9393a61ef724e698ab8d325a" ESC[37mDEBUESC[0m[5863] containerd: process exited                    ESC[37midESC[0m=75190c7ed862395af5a4dfed025f6f27a3bb39da9393a61ef724e698ab8d325a ESC[37mpidESC[0m=init ESC[37mstatusESC[0m=137 ESC[37msystemPidESC[0m=9973
time="2016-10-05T13:59:38.715182786Z" level=debug msg="libcontainerd: received containerd event: &types.Event{Type:\"exit\", Id:\"75190c7ed862395af5a4dfed025f6f27a3bb39da9393a61ef724e698ab8d325a\", Status:0x89, Pid:\"init\", Timestamp:(*timestamp.Timestamp)(0xc820772e00)}" 
time="2016-10-05T13:59:38.715501347Z" level=debug msg="Revoking external connectivity on endpoint pedantic_franklin (57dd9c924af5b9c48f9d193b748aa57dc8efb101786132ecdc1c55842979ec8c)" 
time="2016-10-05T13:59:38.756258676Z" level=debug msg="Releasing addresses for endpoint pedantic_franklin's interface on network bridge" 
time="2016-10-05T13:59:38.756291308Z" level=debug msg="ReleaseAddress(LocalDefault/10.42.0.0/16, 10.42.0.2)" 
time="2016-10-05T13:59:39.938467748Z" level=debug msg="Calling DELETE /v1.24/containers/75190c7ed862395af5a4dfed025f6f27a3bb39da9393a61ef724e698ab8d325a?force=1" 

time="2016-10-05T14:01:56.270713926Z" level=debug msg="clean 182 unused exec commands" 
time="2016-10-05T14:06:56.270678517Z" level=debug msg="clean 152 unused exec commands" 
```

Log output of a "docker stop" called from the Jenkins plugin when the daemon does not respond afterwards:
```
time="2016-10-05T14:59:33.611998832Z" level=debug msg="Calling POST /v1.24/containers/b78f9f03b798f25a9eee84ac9db53f20bb7a6edf8059b640846c128b9f1f1e59/stop?t=10" 
<< nothing, after calling stop, follows >>
```

After such a container hangs it is possible to call "docker run" but not "docker ps" or "docker kill b78f9f03b798f25a9eee84ac9db53f20bb7a6edf8059b640846c128b9f1f1e59" to get rid of the hanging container.

### LINKED ISSUE #27607: docker ps hangs and does not log on docker 1.12.2
**Description**
We run many services with docker and kubernetes in a cluster. On upgrading to docker 1.12.2, we saw that `docker ps` on about 25% of the hosts was hanging. docker also stops logging when this is occurring, and we see error logs in the kubelet of the affected host. The goroutine stack traces I have attached below should have the most relevant info to diagnosing this issue.

**Steps to reproduce the issue:**
1. start a kube cluster on bare metal running kube 1.4.0 and docker 1.12.2
2. start up various services to get a workload
3. check the docker logs and/or run `docker ps` on the hosts (masters and nodes)

**Describe the results you received:**
docker ps hangs, the logs stop logging on about 25% of hosts

**Additional information you deem important (e.g. issue happens only occasionally):**
- docker run is still functional when this is occurring (as are docker info, version, etc.)
- the kubernetes kubelet logs with `GenericPLEG: Unable to retrieve pods: operation timeout: context deadline exceeded`
- restarting the docker daemon resolves the issue, although we have seen it recur later on the same host
- running `strace -fF -p <pid>` on the docker daemon pid also resolves the issue, and causes the logs to be flooded with error messages like `Handler for GET /containers/json returned error: write unix /var/run/docker.sock->@: write: broken pipe`. See the attached docker.log for the full output (~8800 lines). [docker_log.txt](https://github.com/docker/docker/files/543063/docker_log.txt)
- getting the goroutine stack trace shows there may be an issue with containers acquiring a semaphore, but it was hard for me to parse. I have attached three different goroutine stack dumps acquired by sending sigusr1.
[goroutine1.txt](https://github.com/docker/docker/files/543066/goroutine1.txt)
[goroutine2.txt](https://github.com/docker/docker/files/543067/goroutine2.txt)
[goroutine3.txt](https://github.com/docker/docker/files/543068/goroutine3.txt)

**Output of `docker version`:**

```
$ docker version
Client:
 Version:      1.12.2
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   bb80604
 Built:        Tue Oct 11 18:15:17 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.2
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   bb80604
 Built:        Tue Oct 11 18:15:17 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**
This is from a host where the issue is currently occuring:
```
$ docker info
Containers: 14
 Running: 12
 Paused: 0
 Stopped: 2
Images: 72
Server Version: 1.12.2
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 175
 Dirperm1 Supported: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: overlay bridge host null
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options:
Kernel Version: 3.13.0-55-generic
Operating System: Ubuntu precise (12.04.4 LTS)
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 1.955 GiB
Name: tuk5m1kmaster1
ID: AAWZ:IW2J:ELCJ:B35B:6YA3:JPBF:3IDG:REKJ:MNYT:NTF4:ANNH:JMKT
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
bare metal
running Ubuntu 12.04 LTS with kernel 3.13.0-55-generic
kubernetes version info:
```Client Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.0", GitCommit:"a16c0a7f71a6f93c7e0f222d961f4675cd97a46b", GitTreeState:"clean", BuildDate:"2016-09-26T18:16:57Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.0", GitCommit:"a16c0a7f71a6f93c7e0f222d961f4675cd97a46b", GitTreeState:"clean", BuildDate:"2016-09-26T18:10:32Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}```

### PR DESCRIPTION
**Pull request by tonistiigi**
fixes #27185
fixes #26561
fixes #27607
- Open stdout/stderr from both sides instead of `O_RDWR` in containerd to wait for client side to be ready.
- Add cancellation of indefinitely blocked fifos
- Better support for closing half-open fifos
- Support closing open fifos that have been deleted

@mlaventure

### CODE REVIEW

**Comment by tonistiigi** [comment]
cc @jhowardmsft for windows changes. I'm not sure why the conversion pipe for stdout/stderr was there so I just replaced it with a wrapper that should be quite faster. Most likely the objects could be used directly, maybe need to manually call close on process exit or error cases.

**Review by lowenna — COMMENTED** [review]
@darrenstahlmsft PTAL

**Comment by runcom** [comment]
@tonistiigi can you update this PR to directly vendor docker/containerd at https://github.com/docker/containerd/commit/52ef1ceb4b660c42cf4ea9013180a5663968d4c7

**Comment by tonistiigi** [comment]
@runcom There seems to be an issue with 3.13 kernel in jenkins where it sometimes gets `ESTALE` on close. I'll update when I have figured it out. Bit hard because can't reproduce outside jenkins.

File: libcontainerd/process_windows.go [hunk 1]
```diff
@@ -18,16 +19,17 @@ type process struct {
...
+type autoClosingReader struct {
+	io.ReadCloser
+	sync.Once
+}
+
+func (r *autoClosingReader) Read(b []byte) (n int, err error) {
+	n, err = r.ReadCloser.Read(b)
+	if err == io.EOF {
+		r.Once.Do(func() { r.ReadCloser.Close() })
+	}
```
Comment (tonistiigi) [inline libcontainerd/process_windows.go]: If you like to. As I commented I'm not sure why this is needed at all. If close needs to be called there are probably better places to do that than detecting if the error from read is fatal or not.

**Comment by tonistiigi** [comment]
@LK4D4 Part of the ci issue seemed to be with closing stdin twice that you already covered with #27391 and also having the close in a goroutine that may cause the directory containing the fifo to be deleted before it is closed. Closing stdin twice can happen when it's closed in `AttachStreams` but after that create/exec still returns an error and fifos are closed for cleanup.

**Comment by thaJeztah** [comment]
ping @LK4D4 @darrenstahlmsft PTAL - is this looking good to you?

**Comment by darrenstahlmsft** [comment]
LGTM

**Comment by mlaventure** [comment]
@tonistiigi this needs a rebase

**Comment by mlaventure** [comment]
LGTM

**Review by thaJeztah — APPROVED** [review]
thanks for reviewing @darrenstahlmsft, LGTM too then

**Comment by runcom** [comment]
Broken master with go < 1.7 (or 1.8 I don't remember where `context` was embedded into golang itself):
```
---> Making bundle: dynbinary (in bundles/1.13.0-dev/dynbinary)
Building: bundles/1.13.0-dev/dynbinary-client/docker-1.13.0-dev
Created binary: bundles/1.13.0-dev/dynbinary-client/docker-1.13.0-dev
Building: bundles/1.13.0-dev/dynbinary-daemon/dockerd-1.13.0-dev
vendor/src/github.com/tonistiigi/fifo/fifo.go:4:2: cannot find package "context" in any of:
        /usr/lib/golang/src/context (from $GOROOT)
        /home/amurdaca/src/github.com/docker/docker/.gopath/src/context (from $GOPATH)
        /home/amurdaca/src/github.com/docker/docker/vendor/src/context
```

**Comment by runcom** [comment]
opened https://github.com/docker/docker/issues/27669

**Pull request merged**